### PR TITLE
feat(sanity_client): add ability to post request

### PR DESCRIPTION
**Pull request by stephenrichter**
This PR adds the ability for the `sanity_client` to do `POST` requests with a GROQ query as the JSON body.

This avoids the issue where a GROQ query string is too long to use with http `GET`.

See https://www.sanity.io/docs/http-query#J0Wj7Tip for more information

### CODE REVIEW

File: packages/sanity/sanity_client/lib/url_builder.dart [hunk 1]
```diff
@@ -132,6 +135,19 @@ final class SanityUrlBuilder extends UrlBuilder<SanityConfig> {
       queryParameters: queryParameters,
     );
   }
+
+  @override
+  Uri postUrl() {
```
Comment (pavanpodila) [inline packages/sanity/sanity_client/lib/url_builder.dart:140]: Why don't we do this with standard query itself by checking for size of the query string. If it's longer we could switch to post method automatically. That will simplify the usage as well. 

Thoughts?

**Review by pavanpodila — CHANGES_REQUESTED** [review]

File: packages/sanity/sanity_client/lib/url_builder.dart (hunk 1 above)
Comment (stephenrichter) [reply packages/sanity/sanity_client/lib/url_builder.dart:140]: @pavanpodila I like this method better. I've added a `SanityRequest` class which handles the logic internally to determine whether the query `requiresPost` or not.

This means the API will remain the same `SanityClient.fetch()` and will do the magic behind the scenes 😉

File: packages/sanity/sanity_client/lib/client.dart [hunk 2]
```diff
@@ -103,17 +103,41 @@ final class SanityClient {
...
-    final response = await httpClient.get(uri, headers: _requestHeaders);
+  Future<SanityQueryResponse> fetch(
+    String query, {
+    Map<String, String>? params,
+  }) async {
+    final request = SanityRequest(
+      urlBuilder: urlBuilder,
+      query: query,
+      params: params,
+    );
```
Comment (pavanpodila) [inline packages/sanity/sanity_client/lib/client.dart:114]: Very nice approach with the request! and neatly captures the essence. I think we can make the SanityRequest private and use the part/part-of directives to ensure its not leaked outside.

Also can we add some tests to ensure coverage does not drop ?

Comment (stephenrichter) [reply packages/sanity/sanity_client/lib/client.dart:115]: It's not recommended to use part/part-of directives, so I've used a `lib/src` folder to maintain `SanityRequest` as private and not expose it with the library. See https://github.com/dart-lang/site-www/issues/2371 for more info.

I've also added test coverage to make sure the 11kb limit is respected and GET/POST are operating correctly.

File: packages/sanity/sanity_client/lib/url_builder.dart (hunk 1 above)
Comment (pavanpodila) [inline packages/sanity/sanity_client/lib/url_builder.dart:140]: can we absorb this as well in the queryUrl() ... I think the choice of url construction is an implementation detail and should not be exposed outside..

**Review by pavanpodila — CHANGES_REQUESTED** [review]

File: packages/sanity/sanity_client/lib/url_builder.dart (hunk 1 above)
Comment (stephenrichter) [reply packages/sanity/sanity_client/lib/url_builder.dart]: Makes sense. Done.

**Review by pavanpodila — APPROVED** [review]

**Comment by pavanpodila** [comment]
Cool PR and implementation @stephenrichter 🥇

**Comment by pavanpodila** [comment]
@all-contributors add @stephenrichter for code,docs,test

**Pull request merged**